core/common/src/internal/util.kt

```diff
@@ -8,3 +8,32 @@ package kotlinx.datetime.internal
 internal fun Char.isAsciiDigit(): Boolean = this in '0'..'9'
 
 internal fun Char.asciiDigitToInt(): Int = this - '0'
+
+/** Working around the JSR-310 behavior of failing to parse long year numbers even when they start with leading zeros */
+private fun removeLeadingZerosFromLongYearForm(input: String, minStringLengthAfterYear: Int): String {
+    // the smallest string where the issue can occur is "+00000002024", its length is 12
+    val failingYearStringLength = 12
+    // happy path: the input is too short or the first character is not a sign, so the year is not in the long form
+    if (input.length < failingYearStringLength + minStringLengthAfterYear || input[0] !in "+-") return input
+    // the year is in the long form, so we need to remove the leading zeros
+    // find the `-` that separates the year from the month
+    val yearEnd = input.indexOf('-', 1).takeIf { it != -1 }
+    // if the year is too short, no need to remove the leading zeros, and if the string is malformed, just leave it
+    if (yearEnd == null || yearEnd < failingYearStringLength) return input
```

Comment:
```suggestion
    val yearEnd = input.indexOf('-', 1)
    // if the year is too short, no need to remove the leading zeros, and if the string is malformed, just leave it
    if (yearEnd < 0 || yearEnd < failingYearStringLength) return input
```

Reply:
Or, in this case the first condition can be removed, though it would be less clear.

Reply:
Good call. With a clarifying comment, it seems fine.